File: 0801-0900/0851-Loud and Rich/0851-Loud and Rich.cpp
class Solution {
public:
    vector<int> loudAndRich(vector<vector<int>>& richer, vector<int>& quiet) {
        int n = quiet.size();
        vector<vector<int>> graph(n);
        for (auto r : richer) {
            graph[r[1]].push_back(r[0]);
        }
        vector<int> result(n, -1);
        for (int i = 0; i < n; ++i) {
            if (result[i] == -1) {
                dfs(graph, i, quiet, result);
            }
        }
        return result;
    }
    
private:
    void dfs(vector<vector<int>>& graph, int curr, vector<int>& quiet, vector<int>& result) {
        result[curr] = curr;
        for (int next : graph[curr]) {
            if (result[next] == -1) {
                dfs(graph, next, quiet, result);                
            }
            if (quiet[result[next]] < quiet[result[curr]]) {
                result[curr] = result[next];
            }
        }        
    }
};

File: 0801-0900/0851-Loud and Rich/0851-Loud and Rich.py
import collections
class Solution:
    def loudAndRich(self, richer: List[List[int]], quiet: List[int]) -> List[int]:
        graph = collections.defaultdict(list)
        for u, v in richer:
            graph[v].append(u)
            
        result = [-1] * len(quiet)
        def dfs(curr):
            result[curr] = curr
            for next in graph[curr]:
                if result[next] == -1:
                    dfs(next)
                if quiet[result[next]] < quiet[result[curr]]:
                    result[curr] = result[next]
        
        for i in range(len(quiet)):
            if result[i] == -1:
                dfs(i)
        return result
